File: src/csharp/core/src/Style/Component.cs
using System.Text.Json.Nodes;

namespace DiceBear.Internal
{
    /// <summary>
    /// Read-only view over an entry in a style definition's <c>components</c>
    /// block.
    /// </summary>
    /// <remarks>
    /// An entry is either a base component with its own dimensions and
    /// variants or an alias declared via <c>extends</c>. Aliases are pure
    /// references — they inherit dimensions, variants and all transforms from
    /// the source.
    /// </remarks>
    internal sealed class Component
    {
        private readonly JsonObject _data;
        private readonly string _name;
        private readonly Component? _source;

        private ComponentTranslate? _translate;
        private OrderedMap<ComponentVariant>? _variants;

        internal Component(string name, JsonObject data, Component? source = null)
        {
            _name = name;
            _data = data;
            _source = source;
        }

        /// <summary>
        /// Returns the entry's own name as declared in the style definition.
        /// For aliases this is the alias key, not the source component's name
        /// (use <see cref="SourceName"/> for the canonical user-option key
        /// prefix).
        /// </summary>
        internal string Name() => _name;

        /// <summary>
        /// Returns the source component name when this entry is an alias, or
        /// <see langword="null"/> for a base component.
        /// </summary>
        internal string? ExtendsName() => JsonRead.Str(_data, "extends");

        /// <summary>
        /// Returns the canonical user-option key prefix: the source
        /// component's name when this entry is an alias, otherwise the entry's
        /// own name.
        /// </summary>
        internal string SourceName() => ExtendsName() ?? _name;

        /// <summary>
        /// Returns the component's intrinsic width in canvas coordinates. For
        /// aliases the source component's width is returned.
        /// </summary>
        internal double Width() => _source is not null ? _source.Width() : JsonRead.Num(_data, "width") ?? 0.0;

        /// <summary>
        /// Returns the component's intrinsic height in canvas coordinates. For
        /// aliases the source component's height is returned.
        /// </summary>
        internal double Height() => _source is not null ? _source.Height() : JsonRead.Num(_data, "height") ?? 0.0;

        /// <summary>
        /// Returns the probability (0–100) that this component is rendered.
        /// Aliases delegate to the source; defaults to 100 (always visible).
        /// </summary>
        internal double Probability() =>
            _source is not null ? _source.Probability() : JsonRead.Num(_data, "probability") ?? 100.0;

        /// <summary>
        /// Returns the rotation range, or <see langword="null"/> when unset.
        /// Aliases delegate to the source.
        /// </summary>
        internal NumberRange? Rotate() => _source is not null ? _source.Rotate() : JsonRead.Range(_data, "rotate");

        /// <summary>
        /// Returns the scale range, or <see langword="null"/> when unset.
        /// Aliases delegate to the source.
        /// </summary>
        internal NumberRange? Scale() => _source is not null ? _source.Scale() : JsonRead.Range(_data, "scale");

        /// <summary>
        /// Returns the translate descriptor. Aliases delegate to the source.
        /// </summary>
        internal ComponentTranslate Translate()
        {
            if (_source is not null)
            {
                return _source.Translate();
            }

            return _translate ??= new ComponentTranslate(
                JsonRead.Obj(_data, "translate") ?? new JsonObject());
        }

        /// <summary>
        /// Returns a name to <see cref="ComponentVariant"/> map for all defined
        /// variants, in definition order. Aliases delegate to the source
        /// component's variants.
        /// </summary>
        internal OrderedMap<ComponentVariant> Variants()
        {
            if (_source is not null)
            {
                return _source.Variants();
            }

            if (_variants is not null)
            {
                return _variants;
            }

            var map = new OrderedMap<ComponentVariant>();
            var variants = JsonRead.Obj(_data, "variants");

            if (variants is not null)
            {
                foreach (var entry in variants)
                {
                    if (entry.Value is JsonObject obj)
                    {
                        map.Set(entry.Key, new ComponentVariant(obj));
                    }
                }
            }

            return _variants = map;
        }
    }
}

File: src/rust/core/src/style/component.rs
//! Read-only view over an entry in a style definition's `components` block.
//!
//! An entry is either a base component with its own dimensions and variants or
//! an alias declared via `extends`. Aliases are pure references: they share the
//! source's [`ComponentData`] via `Arc` and delegate every accessor to it.

use std::collections::HashMap;
use std::sync::Arc;

use indexmap::IndexMap;
use serde::Deserialize;

use super::component_translate::ComponentTranslate;
use super::component_variant::ComponentVariant;
use crate::prng::Range;

/// A `components` entry before alias resolution.
#[derive(Deserialize)]
#[serde(untagged)]
pub(crate) enum RawComponent {
    Alias { extends: String },
    // Boxed: the base variant is far larger than the alias.
    Base(Box<ComponentData>),
}

/// The intrinsic data of a base component, shared by its aliases via `Arc`.
#[derive(Deserialize)]
pub(crate) struct ComponentData {
    width: f64,
    height: f64,
    probability: Option<f64>,
    rotate: Option<Range>,
    scale: Option<Range>,
    #[serde(default)]
    translate: ComponentTranslate,
    variants: IndexMap<String, ComponentVariant>,
}

impl ComponentData {
    /// The variants of a base component, before alias resolution. Consumed by
    /// the definition-level walks in `Style` (animation validation and
    /// detection).
    pub(crate) fn variants(&self) -> &IndexMap<String, ComponentVariant> {
        &self.variants
    }
}

pub(crate) struct Component {
    name: String,
    source_name: String,
    data: Arc<ComponentData>,
}

impl Component {
    /// The entry's own name (the alias key for aliases).
    pub(crate) fn name(&self) -> &str {
        &self.name
    }

    /// The canonical user-option key prefix: the source's name for an alias,
    /// otherwise the entry's own name.
    pub(crate) fn source_name(&self) -> &str {
        &self.source_name
    }

    /// Whether this entry is an alias (its name differs from its source's).
    pub(crate) fn is_alias(&self) -> bool {
        self.name != self.source_name
    }

    pub(crate) fn width(&self) -> f64 {
        self.data.width
    }

    pub(crate) fn height(&self) -> f64 {
        self.data.height
    }

    /// The render probability (0–100), defaulting to `100`.
    pub(crate) fn probability(&self) -> f64 {
        self.data.probability.unwrap_or(100.0)
    }

    pub(crate) fn rotate(&self) -> Option<Range> {
        self.data.rotate
    }

    pub(crate) fn scale(&self) -> Option<Range> {
        self.data.scale
    }

    pub(crate) fn translate(&self) -> &ComponentTranslate {
        &self.data.translate
    }

    pub(crate) fn variants(&self) -> &IndexMap<String, ComponentVariant> {
        &self.data.variants
    }
}

/// Resolves the raw `components` block into flattened [`Component`]s. Base
/// components are built first; each alias then shares its source's `Arc`. An
/// alias whose source is missing is skipped (alias validity is checked in
/// `Style` so a clear error is reported instead).
pub(crate) fn build(raw: Option<IndexMap<String, RawComponent>>) -> IndexMap<String, Component> {
    let Some(raw) = raw else {
        return IndexMap::new();
    };

    let mut bases: HashMap<String, Arc<ComponentData>> = HashMap::new();
    let mut order: Vec<(String, Option<String>)> = Vec::new();

    for (name, entry) in raw {
        match entry {
            RawComponent::Base(data) => {
                bases.insert(name.clone(), Arc::new(*data));
                order.push((name, None));
            }
            RawComponent::Alias { extends } => order.push((name, Some(extends))),
        }
    }

    let mut components = IndexMap::new();

    for (name, extends) in order {
        let key = extends.as_deref().unwrap_or(&name);

        if let Some(data) = bases.get(key) {
            let source_name = extends.clone().unwrap_or_else(|| name.clone());
            components.insert(
                name.clone(),
                Component {
                    name,
                    source_name,
                    data: Arc::clone(data),
                },
            );
        }
    }

    components
}
